### PR TITLE
fix(engine-server): do not escape `<style>` content

### LINKED ISSUE #3439: `<style>` text contents are improperly escaped in SSR
In SSR, selectors like `[data-foo="bar"]` become `[data-foo=&quot;bar&quot;]`, which is not valid CSS and thus fails to render styles correctly.

Repro: 10fff805b27121b3f08b5f08ea556b6edd38ad0d

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

Fixes #3439

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ⚠️ Yes, it does include an observable change.

If you are using `</style>` in your CSS anywhere, then it will throw an error during SSR. This is probably a good thing, as we don't want to be vulnerable to XSS attacks.

I also can't imagine why someone would want to put a `</style>` in their CSS (as a comment maybe?). If someone does, and we need to fix that bug, then we can cross that bridge when we get there.

## GUS work item
[W-12949456](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001ONT5IYAX/view)

### CODE REVIEW

File: packages/@lwc/engine-server/src/__tests__/fixtures.spec.ts [hunk 1]
```diff
@@ -59,69 +58,12 @@ async function compileFixture({ input, dirname }: { input: string; dirname: stri
...
-            res += getPadding() + src.slice(start, pos) + '\n';
-        }
-    }
-
-    return res.trim();
+    return format(src, {
+        parser: 'html',
+        htmlWhitespaceSensitivity: 'ignore',
+        bracketSameLine: true,
+    });
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures.spec.ts:66]: This is a nice idea, but I am not smart enough to maintain my own HTML parser/formatter. I could not figure out how to modify this to correctly handle unescaped characters like `>` and `<` inside of `<style>` tags.

File: packages/@lwc/engine-server/src/__tests__/fixtures/attribute-aria-modify/expected.html [hunk 2]
```diff
@@ -1,4 +1,3 @@
 <x-cmp aria-busy="true" aria-activedescendant="foo">
-  <template shadowroot="open">
-  </template>
-</x-cmp>
\ No newline at end of file
+  <template shadowroot="open"></template>
+</x-cmp>
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures/attribute-aria-modify/expected.html:3]: Switching to prettier means I am modifying a bazillion fixture files, but the only new ones are `styles-escaping-unsafe` and `styles-escaping-safe`.

File: packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts:1]: There is probably a super-clever way to detect that the contents of a `<style>` tag are going to result in an XSS vulnerability, but the most bulletproof solution I could think of is to parse the whole thing in parse5 and see if it results in the expected single `<style>` tag.

There is probably also a way to optimize this (e.g. check that the string actually contains any special characters), but again, the most bulletproof thing is to just check everything. If this turns out to be a perf problem later, then we can optimize it then.

File: packages/@lwc/engine-server/src/__tests__/fixtures/attribute-boolean/expected.html [hunk 3]
```diff
@@ -1,5 +1,5 @@
 <x-attribute-boolean>
   <template shadowroot="open">
-    <input type="checkbox" required readonly checked>
+    <input type="checkbox" required readonly checked />
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures/attribute-boolean/expected.html:3]: I'm not super happy about this (`input` is self-closing), but Prettier doesn't seem to have an option to turn this off.

File: packages/@lwc/engine-server/src/__tests__/fixtures/styles-escaping-safe/expected.html [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<x-styles>
+  <template shadowroot="open">
+    <style type="text/css">
+      [data-foo="'&
+      <>
+        "] {color: black;}[data-foo='"'] {color: white;}
+      </style>
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures/styles-escaping-safe/expected.html:7]: I'm still trying to figure out how to fix the fixture formatter. Just trust me that the serialized HTML doesn't really look like this.

Comment (nolanlawson) [reply packages/@lwc/engine-server/src/__tests__/fixtures/styles-escaping-safe/expected.html:7]: I've struggled with this for quite some time. Other options (e.g. switching to Prettier) had their own downsides. I think for now I'm fine with just having the fixture look a little ugly.

Comment (nolanlawson) [reply packages/@lwc/engine-server/src/__tests__/fixtures/styles-escaping-safe/expected.html]: Fixed it

File: packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const style = node.childNodes[0];
+
+    if (style.nodeName !== 'style' || style.childNodes.length !== 1) {
+        return false;
+    }
+
+    const textNode = style.childNodes[0];
+
+    return textNode.nodeName === '#text';
+}
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts:24]: There is probably a super-clever way to detect that the contents of a `<style>` tag are going to result in an XSS vulnerability, but the most bulletproof solution I could think of is to parse the whole thing in parse5 and see if it results in the expected single `<style>` tag.

There is probably also a way to optimize this (e.g. check that the string actually contains any special characters), but again, the most bulletproof thing is to just check everything. If this turns out to be a perf problem later, then we can optimize it then.

Comment (jye-sf) [reply packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts:24]: parse5 doesn't have built-in sanitization right? We're relying on
```
(single <style> tag + single text node) === SAFE
```
?

Comment (nolanlawson) [reply packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts:24]: Exactly, yes. This is why I have a test where the CSS file contains `</style>`. I can add a few more to catch a few more scenarios as well.

Comment (nolanlawson) [reply packages/@lwc/engine-server/src/utils/validate-style-text-contents.ts:24]: Done: d5242e75d

File: packages/@lwc/engine-server/package.json [hunk 6]
```diff
@@ -26,12 +26,14 @@
     "//": {
         "parse5": "Currently can't upgrade parse5: dropped cjs support in v7.0.0 https://github.com/inikulin/parse5/releases/tag/v7.0.0"
     },
+    "dependencies": {
+        "parse5": "^6.0.1"
+    },
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/package.json:31]: parse5 is now used inside of `@lwc/engine-server` at runtime (not just during testing), so it now needs to be a regular dep, not a devDep.

**Review by jye-sf — APPROVED** [review]

**Comment by nolanlawson** [comment]
I would love to get one extra +1 just to play it safe. /cc @pmdartus @divmain

**Comment by divmain** [comment]
This seems sane to me. One question: what do you think about moving the validation to compilation? As is, this change wil incur a small cost for every SSR of every component (that has styles). If it were moved to compilation, the validation would only occur once, and style content could be trusted and left unescaped during serialization.

**Comment by nolanlawson** [comment]
@divmain That's a great idea, but unfortunately, if we move it to compilation, then it could potentially break for existing LEX components that (for whatever reason) have the string `</style>` in their CSS. As-is, this PR can only possibly break SSR users, not LEX users.

I can open a tracking issue to move the validation to compilation. Sound good?

**Comment by divmain** [comment]
That works for me!

**Review by divmain — APPROVED** [review]

**Pull request merged**